File: android/swjtu_xszc/app/src/main/java/com/jtutang/swjtu_xszc/SwjtuXszc.java
package com.jtutang.swjtu_xszc;

import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by jtu唐 on 2015/5/25.
 */
public class SwjtuXszc {
    private String username;
    private String password;
    private String cookie;
    private String loginUrl = "http://ocw.swjtu.edu.cn/servlet/UserLoginDataAction";
    private String checkUrl = "http://ocw.swjtu.edu.cn/websys/studentstudytime.jsp";
    private String listUrl = "http://ocw.swjtu.edu.cn/page/viewVideo.jsp?c_id=196";
    private String Referer = "http://ocw.swjtu.edu.cn/websys/videoview.jsp?resource_id=";
    private String sid;
    private String sourceID;
    public SwjtuXszc(String username,String password){
        if(username.isEmpty() || password.isEmpty())return;
        this.username = username;
        this.password = password;
    }
    public boolean isLogin(){
        String restr = null;
        try{
            URL url = new URL(this.checkUrl);
            HttpURLConnection connection = (HttpURLConnection)url.openConnection();
            connection.setConnectTimeout(5000);
            connection.setRequestProperty("Cookie", this.cookie);
            InputStreamReader is = new InputStreamReader(connection.getInputStream());
            BufferedReader br = new BufferedReader(is);
            String line = "";
            while((line = br.readLine()) != null) {
                restr += line;
            }
            connection.disconnect();
            is.close();
        }catch (Exception e){
        }
        Log.e("DASHABI",restr);
        if(restr.indexOf("学时") >= 0)return true;
        return false;
    }

    public void chooseVideo(ArrayList<String> al){
        int len = al.size() - 1;
        int rand = (int)Math.round(Math.random()*len);
        this.sourceID = al.get(rand).split("=")[1];
    }

    public boolean login(){
        String restr = "";
        try{
            String key = null;
            String cookieVal = null;
            URL url = new URL(this.loginUrl);
            String data = "UserName=" + this.username + "&Password=" + this.password + "&UserType=stu";
            HttpURLConnection connection = (HttpURLConnection)url.openConnection();
            connection.setDoOutput(true);
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
            connection.setConnectTimeout(5000);
            OutputStream os = connection.getOutputStream();
            os.write(data.getBytes());
            InputStreamReader is = new InputStreamReader(connection.getInputStream());
            BufferedReader br = new BufferedReader(is);
            for (int i = 1; (key = connection.getHeaderFieldKey(i)) != null; i++ ) {
                if (key.equalsIgnoreCase("set-cookie")) {
                    cookieVal = connection.getHeaderField(i);
                    cookieVal = cookieVal.substring(0, cookieVal.indexOf(";"));
                    this.cookie = cookieVal + ";";
                }
            }
            String line = "";
            while((line = br.readLine()) != null){
                restr += line;
            }
            connection.disconnect();
            is.close();
        }catch (IOException e){
        }
        Log.e("DASHABI",restr);
        if(restr.indexOf("成功") >= 0)return true;
        return false;
    }

    public ArrayList<String> grap_list(){
        String restr = null;
        ArrayList<String> relist = new ArrayList<String>();
        try{
            URL url = new URL(this.listUrl);
            HttpURLConnection connection = (HttpURLConnection)url.openConnection();
            connection.setConnectTimeout(5000);
            InputStreamReader is = new InputStreamReader(connection.getInputStream());
            BufferedReader br = new BufferedReader(is);
            String line = "";
            while((line = br.readLine()) != null) {
                restr += line;
            }
            connection.disconnect();
            is.close();
        }catch (Exception e){
        }
        Pattern p = Pattern.compile("<a href=\"(.*?)\" target=_blank>");
        Matcher m = p.matcher(restr);
        while(m.find()){
            relist.add(m.group(1));
        }
        Log.e("DASHABI",restr);
        return relist;
    }

    public String cheat(String status){
        String restr = null;
        String utf8Restr = null;
        try{
            String Url = "http://ocw.swjtu.edu.cn/servlet/UserStudyRecordAction?resource_id=" + this.sourceID;
            Url += "&SetType=" + status;
            if("ADD" == status) Url += "&ranstring=&sid=";
            else Url += "&ranstring=&sid=" + this.sid;
            Url += "&tt=" + System.currentTimeMillis();
            Log.e("DASHABI",Url);
            String referer = this.Referer + this.sourceID;
            URL url = new URL(Url);
            HttpURLConnection connection = (HttpURLConnection)url.openConnection();
            connection.setConnectTimeout(5000);
            connection.setRequestProperty("Cookie", this.cookie);
            connection.setRequestProperty("User-Agent","User-Agent:Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/38.0.2125.111 Safari/537.36");
            connection.setRequestProperty("Referer",referer);
            InputStreamReader is = new InputStreamReader(connection.getInputStream(),"GBK");
            BufferedReader br = new BufferedReader(is);
            String line = "";
            while((line = br.readLine()) != null) {
                restr += line;
            }
            connection.disconnect();
            is.close();
            utf8Restr = new String(restr.getBytes("UTF-8"),"ISO-8859-1");
            utf8Restr = new String(utf8Restr.getBytes("ISO-8859-1"),"UTF-8");
        }catch (Exception e){
        }
        Log.e("DASHABI",utf8Restr);
        Pattern p = Pattern.compile("<select_message>(.*?)<\\/select_message>");
        Matcher m = p.matcher(utf8Restr);
        if("ADD" == status && m.find()){
            this.sid = m.group(1);
            return m.group(1);
        }
        if(utf8Restr.indexOf("结束") >= 0)return "success";
        return "fail";
    }
}

File: python/swjtu_xszc.py
#encoding:utf-8
import urllib2
import cookielib
import re
import random
import time
#sjwtu刷形势政策脚本
#Author	jtu唐
#QQ		461881920
#邮箱	jtutang@163.com
class swjtu_xszc():
	#登录请求链接
	login_url 	= "http://ocw.swjtu.edu.cn/servlet/UserLoginDataAction"
	#检测登录状态的链接
	check_url 	= "http://ocw.swjtu.edu.cn/websys/studentstudytime.jsp"
	#视频列表链接
	list_url 	= "http://ocw.swjtu.edu.cn/page/viewVideo.jsp?c_id=196"
	def __init__(self,username,password):
		self.username = username
		self.password = password

	#登录函数
	#登录成功返回True,否则返回False
	def login(self):
		param = "UserName=%s&Password=%s&UserType=stu" % (self.username,self.password)
		self.cj = cookielib.CookieJar()
		self.opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.cj))
		res = self.opener.open(self.login_url,param).read()
		return "登录成功" in res
	
	#抓取视频链接函数
	#初始化后应首先运行此函数
	def grap_list(self):
		info = urllib2.urlopen(self.list_url).read()
		lr = re.compile(r'<a href="(.*?)" target=_blank>')
		return re.findall(lr,info)

	#
	#
	def choose_Video(self,video_list):
		self.source_id = random.sample(video_list,1)[0].split("=")[1]
		return self.source_id

	#检测登录状态的函数
	#登录状态存在返回True,否则返回False
	def check_login(self):
		res = self.opener.open(self.check_url).read()
		return "计时统计" in res

	#*******************
	#	欺骗函数
	#*******************
	#
	#@param status 要欺骗的状态  开始记时为'ADD',停止记时为'STOP'
	#
	def cheat(self,status):
		referer = "http://ocw.swjtu.edu.cn/websys/videoview.jsp?resource_id=%s" % self.source_id
		time_url = "http://ocw.swjtu.edu.cn/servlet/UserStudyRecordAction?resource_id=%s" % self.source_id
		time_url += "&SetType=%s&ranstring=&" % status
		if "STOP" == status:
			time_url += "sid=%s&tt=%s" % (self.sid,str(int(time.time())))
		time_url += "sid=&tt=%s" % str(int(time.time()))
		request = urllib2.Request(time_url)
		request.add_header("Referer",referer)
		request.add_header("User-Agent","ozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/38.0.2125.111 Safari/537.36")
		rs = self.opener.open(request).read()
		sr = re.compile(r'<select_message>(.*?)</select_message>')
		smg = re.findall(sr,rs)[0]
		if "ADD" == status:
			self.sid = smg
		return smg